src/components/cache/managers/ioredis/ioredis-cache.manager.ts

```diff
@@ -0,0 +1,78 @@
+import type { Redis } from 'ioredis';
+import { PackageUtils } from '../../../../utils/package-loader';
+import { PrefixedManager } from '../prefixed-manager.abstract';
+import {
+  CacheValue,
+  ICacheManager,
+  ICacheManagerCollection,
+  ICacheManagerMap,
+  SetOptions,
+} from '../cache.manager.interface';
+import { IORedisCacheMap } from './ioredis-cache.map';
+import { IORedisCacheCollection } from './ioredis-cache.collection';
+import { ICacheValueSerializer } from '../../serializers/types';
+import { JsonSerializer } from '../../serializers/json.serializer';
+import type { RedisOptions } from 'ioredis';
+
+export type IIORedisOptions = RedisOptions;
+
+export class IORedisCacheManager<T extends CacheValue> extends PrefixedManager implements ICacheManager<T> {
+  private readonly serializer: ICacheValueSerializer;
+
+  private constructor(private readonly redisManager: Redis, prefix = '') {
+    super(prefix);
+
+    this.serializer = new JsonSerializer();
+  }
+
+  static async create<T extends CacheValue>(options?: IIORedisOptions, prefix = ''): Promise<IORedisCacheManager<T>> {
+    const RedisCtor = PackageUtils.loadPackage<any>('ioredis');
+
+    return new IORedisCacheManager<T>(new RedisCtor(options), prefix);
+  }
+
+  public async set<V extends T>(key: string, data: V, options?: SetOptions): Promise<void> {
+    if (options?.expiresInSeconds) {
+      await this.redisManager.set(this.withPrefix(key), JSON.stringify(data), 'EX', options.expiresInSeconds);
+    } else {
+      await this.redisManager.set(this.withPrefix(key), JSON.stringify(data));
+    }
+  }
+
+  public async get<V extends T>(key: string): Promise<V | null> {
+    const stringifiedData = await this.redisManager.get(this.withPrefix(key));
+    return stringifiedData ? JSON.parse(stringifiedData) : null;
+  }
+
+  public async del(key: string[]): Promise<void> {
+    if (key.length) {
+      await this.redisManager.del(key.map(this.withPrefix.bind(this)));
+    }
+  }
+
+  async expire(keys: string[], ttlMs: number): Promise<void> {
+    for (const key of keys) {
+      await this.redisManager.pexpire(this.withPrefix(key), ttlMs);
+    }
+  }
+
+  forScope<S extends CacheValue>(prefix?: string): ICacheManager<S> {
```

Comment:
Well, no. The intention was to create the new (scoped) instance of the same class, using the very same Redis connection (or NodeCache instance in case of `local`), but with different prefix and (important) - different typing.

When you use `FronteggCache.getInstance()` static method, in return you're receiving the instance and you don't really know (or - don't really NEED TO KNOW) what class it is. It's just fulfilling the `ICacheManager` interface. If this method is static, you'd need to iterate over known implementation and check which instance it is - and that's an antipattern, because in your code you need to know about the `ICacheManager`, not about particular implementations.

Usage of non-static method allows to apply the polymorphism here, so cache implementations know how to actually produce their scoped copies and don' leak implementation details that other parts of code should not be concerned about.